File: py/openaxis/tests/test_session.py
"""Shared event traces for the private Navigation session prototype."""
import json
from pathlib import Path

import pytest

from openaxis._session import SessionState
from openaxis.types import CameraPose


FIXTURE = Path(__file__).resolve().parents[3] / "fixtures/openaxis-1.0/session.json"
SCENARIOS = json.loads(FIXTURE.read_text(encoding="utf-8"))["scenarios"]


def pose(value):
    if value is None:
        return None
    if isinstance(value, (int, float)):
        value = {"x": value}
    return CameraPose(t=tuple(value.get("t", [value.get("x", 0), 0, 0])),
                      r=tuple(value.get("r", [0, 0, 0])),
                      fov=None if "extent" in value else value.get("fov", 1),
                      ortho_extent=value.get("extent"))


@pytest.mark.parametrize("scenario", SCENARIOS, ids=lambda case: case["name"])
def test_shared_session_trace(scenario):
    state = SessionState()
    tokens, tickets, writes = {}, {}, {}
    for index, event in enumerate(scenario["events"]):
        op = event["op"]
        token = tokens[event["token"]] if "token" in event else state.token
        actual = pose(event.get("actual"))
        now = event.get("now", 0)
        effect = None
        kind = "ok"
        if op == "connection":
            state.connection()
        elif op == "start":
            tokens[event.get("as", "last")] = state.start(event["gesture"])
        elif op == "query":
            ok = state.camera_query(token, actual, scoped=event.get("scoped", True),
                                    supplied=event.get("supplied", True))
            kind = "ok" if ok else "reject"
        elif op == "receive":
            value = pose(event["pose"])
            value = CameraPose(**{**value.__dict__, "gesture_id": event.get("gesture", state.gesture_id),
                                  "seq": event["seq"], "applied_delta_id": event.get("ack")})
            ticket = state.receive(event.get("epoch", state.epoch), value)
            kind = "accepted" if ticket is not None else "reject"
            if ticket is not None:
                tickets[event.get("as", "last")] = ticket
        elif op == "process":
            effect = state.process(tickets[event.get("ticket", "last")], actual, now)
        elif op == "observe":
            effect = state.observe(token, actual, now)
        elif op == "complete":
            effect = state.complete_write(writes[event.get("write", "last")], actual, now,
                                          success=event.get("success", True))
        elif op == "end":
            kind = "ok" if state.end(token) else "reject"
        elif op == "finish":
            kind = "ok" if state.finish(token) else "reject"
        elif op == "cancel":
            effect = state.cancel(token, event["reason"])
        elif op == "timeout":
            effect = state.expire(token, event["delta"], now)
        elif op == "send_failed":
            effect = state.send_failed(token, event["delta"])
        else:
            raise AssertionError(f"Unknown trace operation: {op}")

        result = dict(kind=effect.kind if effect else kind,
                      baseline=state.baseline.t[0] if state.baseline else None,
                      ready=state.ready, pending=state.pending_id, active=state.gesture_id,
                      received=state.last_received, applied=state.last_applied)
        if effect:
            result.update(delta_id=effect.delta_id, gesture_id=effect.gesture_id, reason=effect.reason)
            if effect.difference:
                result.update(t=list(effect.difference.t), r=list(effect.difference.r), scale=effect.difference.scale)
            if effect.write:
                writes[event.get("as", "last")] = effect.write
        for key, expected in event["expect"].items():
            observed = result[key]
            label = f"event {index}: {op}, field {key}"
            if isinstance(expected, (float, list)):
                assert observed == pytest.approx(expected, abs=1e-8), label
            else:
                assert observed == expected, label


def test_private_session_not_exported_at_package_root():
    import openaxis
    assert not hasattr(openaxis, "SessionState")

File: cpp/tests/test_session.cpp
#include "../src/session_state.hpp"
#include <fstream>
#include <iostream>
using namespace openaxis;
using namespace openaxis::detail;
bool matches(const Value &got, const Value &expected) {
    if (got.is_number() && expected.is_number())
        return std::abs(got.get<double>() - expected.get<double>()) < 1e-8;
    if (got.is_array() && expected.is_array()) {
        if (got.size() != expected.size())
            return false;
        for (std::size_t i = 0; i < got.size(); ++i)
            if (!matches(got[i], expected[i]))
                return false;
        return true;
    }
    return got == expected;
}
std::optional<Pose> pose(Value v) {
    if (v.is_null())
        return {};
    if (v.is_number())
        v = Value{{"x", v}};
    Pose p;
    p.t = vector_from(v.value("t", Value{v.value("x", 0.), 0, 0}));
    p.r = vector_from(v.value("r", Value{0, 0, 0}));
    p.fov = v.contains("extent") ? 0 : v.value("fov", 1.);
    p.ortho_extent = v.value("extent", 0.);
    return p;
}
int main() try {
    std::ifstream f(std::string(OPENAXIS_FIXTURES) + "/session.json");
    Value fixture;
    f >> fixture;
    for (auto &scenario : fixture["scenarios"]) {
        SessionState s;
        std::map<std::string, Token> tokens;
        std::map<std::string, Accepted> tickets;
        std::map<std::string, std::shared_ptr<Write>> writes;
        int index = 0;
        for (auto &v : scenario["events"]) {
            std::string op = v["op"], name = v.value("as", "last");
            auto token = v.contains("token") ? tokens.at(v["token"]) : s.token();
            auto actual = pose(v.value("actual", Value{}));
            double now = v.value("now", 0.);
            Effect e;
            bool effect = false;
            std::string kind = "ok";
            if (op == "connection")
                s.connection();
            else if (op == "start")
                tokens[name] = s.start(v["gesture"]);
            else if (op == "query")
                kind = s.query(token, actual, v.value("scoped", true), v.value("supplied", true))
                           ? "ok"
                           : "reject";
            else if (op == "receive") {
                auto ticket = s.receive(v.value("epoch", s.epoch), v.value("gesture", s.gesture.value_or(-1)),
                                        v["seq"], *pose(v["pose"]),
                                        v.contains("ack") ? std::optional<std::int64_t>(v["ack"])
                                                          : std::nullopt);
                kind = ticket ? "accepted" : "reject";
                if (ticket)
                    tickets.insert_or_assign(name, *ticket);
            } else if (op == "end")
                kind = s.end(token) ? "ok" : "reject";
            else if (op == "finish")
                kind = s.finish(token) ? "ok" : "reject";
            else {
                effect = true;
                if (op == "process")
                    e = s.process(tickets.at(v.value("ticket", "last")), actual, now);
                else if (op == "observe")
                    e = s.observe(token, actual, now);
                else if (op == "complete")
                    e = s.complete(writes.at(v.value("write", "last")), actual, now,
                                   v.value("success", true));
                else if (op == "cancel")
                    e = s.cancel(token, v["reason"]);
                else if (op == "timeout")
                    e = s.expire(token, v["delta"], now);
                else if (op == "send_failed")
                    e = s.send_failed(token, v["delta"]);
                else
                    throw std::runtime_error("unknown op");
            }
            Value result = {{"kind", effect ? e.kind : kind},
                            {"baseline", s.baseline ? Value(s.baseline->t.x) : Value{}},
                            {"ready", s.ready},
                            {"pending", s.pending ? Value(*s.pending) : Value{}},
                            {"active", s.gesture ? Value(*s.gesture) : Value{}},
                            {"received", s.received},
                            {"applied", s.applied},
                            {"delta_id", e.delta ? Value(*e.delta) : Value{}},
                            {"gesture_id", e.gesture ? Value(*e.gesture) : Value{}},
                            {"reason", e.reason.empty() ? Value{} : Value(e.reason)}};
            if (e.difference) {
                auto d = *e.difference;
                result["t"] = vector_value(d.t);
                result["r"] = vector_value(d.r);
                result["scale"] = std::abs(d.scale - 1) > 1e-7 ? Value(d.scale) : Value{};
            }
            if (e.write)
                writes[name] = e.write;
            for (auto it = v["expect"].begin(); it != v["expect"].end(); ++it) {
                auto got = result.at(it.key());
                if (!matches(got, it.value()))
                    throw std::runtime_error(scenario["name"].get<std::string>() + " event " +
                                             std::to_string(index) + " " + it.key() + " got " +
                                             got.dump() + " expected " + it.value().dump());
            }
            ++index;
        }
    }
    std::cout << fixture["scenarios"].size() << " shared session traces passed\n";
} catch (const std::exception &e) {
    std::cerr << e.what() << '\n';
    return 1;
}
